### PR TITLE
Switch to CMake-based build

### PR DESCRIPTION
**Pull request by mthrok**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+project(torchaudio)
+
+option(BUILD_SOX "Build SoX and bind statically" OFF)
+option(BUILD_LIBTORCHAUDIO "Build C++ Library" ON)
+option(BUILD_PYTHON_EXTENSION "Build Python extension" OFF)
+
+set(CMAKE_CXX_STANDARD 14 CACHE STRING "The C++ standard whose features are requested to build this target.")
+
+find_package(Torch REQUIRED)
+# TORCH_CXX_FLAGS contains proper _GLIBCXX_USE_CXX11_ABI
```
Comment (cpuhrsch) [inline CMakeLists.txt:12]: You can add a reference to the issue and a bit more color

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+# We propagate it to CMAKE_CXX_FLAGS so that third party libraries are built
+# with the same flag
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${TORCH_CXX_FLAGS}")
+
+add_subdirectory(third_party)
+
+################################################################################
+# Begining of torchaudio extension configuration
```
Comment (cpuhrsch) [inline CMakeLists.txt:20]: nit: Beginning

File: third_party/CMakeLists.txt [hunk 3]
```diff
@@ -1,82 +1,165 @@
...
+  target_link_libraries(opusfile INTERFACE ${INSTALL_DIR}/lib/libopusfile.a)
+  target_include_directories(opusfile INTERFACE ${INSTALL_DIR}/include)
+
+  ExternalProject_Add(sox_
+    PREFIX ${CMAKE_CURRENT_SOURCE_DIR}
+    DEPENDS ogg_ flac_ vorbis_ opusfile_ mp3lame_ mad_
+    DOWNLOAD_DIR ${ARCHIVE_DIR}
+    URL https://downloads.sourceforge.net/project/sox/sox/14.4.2/sox-14.4.2.tar.bz2
+    URL_HASH SHA256=81a6956d4330e75b5827316e44ae381e6f1e8928003c6aa45896da9041ea149c
+    CONFIGURE_COMMAND ${CMAKE_COMMAND} -E env ${envs} ${CMAKE_CURRENT_SOURCE_DIR}/src/sox_/configure ${COMMON_ARGS} --with-lame --with-flac --with-mad --with-oggvorbis --without-alsa --without-coreaudio --without-png --without-oss --without-sndfile --with-opus --disable-openmp
```
Comment (vincentqb) [inline third_party/CMakeLists.txt:138]: nit: comment about `disable-openmp` removed?

File: third_party/CMakeLists.txt [hunk 4]
```diff
@@ -1,82 +1,165 @@
...
+
+  include(ExternalProject)
+
+  set(INSTALL_DIR ${CMAKE_CURRENT_SOURCE_DIR}/install)
+  set(ARCHIVE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/archives)
+  set(COMMON_ARGS --quiet --disable-shared --enable-static --prefix=${INSTALL_DIR} --with-pic --disable-dependency-tracking --disable-debug --disable-examples --disable-doc)
+  set(envs "PKG_CONFIG_PATH=${INSTALL_DIR}/lib/pkgconfig" "LDFLAGS=-L${INSTALL_DIR}/lib $ENV{LDFLAGS}" "CPPFLAGS=-I${INSTALL_DIR}/include $ENV{CPPFLAGS}" "CXXFLAGS=${TORCH_CXX_FLAGS} $ENV{CXXFLAGS}")
+
+
+  ExternalProject_Add(mad_
```
Comment (cpuhrsch) [inline third_party/CMakeLists.txt:13]: As a follow-up, we could eventually use submodules

**Comment by mthrok** [comment]
Closing as the commit history gets very convoluted

**Pull request closed without merging**